### PR TITLE
Upgrade: Improve heuristics around important codemod

### PR DESCRIPTION
**Pull request by philipp-spiess**
This PR improves the heuristics around the important codemod (e.g. `!border` => `border!`) as we noticed a few more cases where we the current heuristics was not enough. 

Specifically, we made it not migrate the candidate in the following conditions:

- When there's an immediate property access: `{ "foo": !border.something + ""}`
- When it's used as condition in the template language: `<div v-if="something && !border"></div>` or `<div x-if="!border"></div>`

## Test plan

I added test cases to the unit tests and updated the integration test to contain a more sophisticated example.

### CODE REVIEW

**Comment by philipp-spiess** [comment]
* **#14774** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14774?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @philipp-spiess and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts [hunk 1]
```diff
@@ -3,6 +3,18 @@ import { parseCandidate } from '../../../../tailwindcss/src/candidate'
 import type { DesignSystem } from '../../../../tailwindcss/src/design-system'
 import { printCandidate } from '../candidates'
 
+const QUOTES = ['"', "'", '`']
+const LOGICAL_OPERATORS = ['&&', '||', '===', '==', '!=', '!==', '>', '>=', '<', '<=']
+const CONDITIONAL_TEMPLATE_SYNTAX = [
+  // Vue
+  /v-if=['"]$/,
+  /v-show=['"]$/,
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/template/codemods/important.ts:11]: Should add `v-else` and `v-else-if` here as well 👍

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts [hunk 2]
```diff
@@ -3,6 +3,18 @@ import { parseCandidate } from '../../../../tailwindcss/src/candidate'
...
+
+  // Alpine
+  /x-show=['"]$/,
+  /x-if=['"]$/,
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/template/codemods/important.ts:15]: Pedantic af but should we put these in the same order as above?

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts [hunk 3]
```diff
@@ -34,32 +46,54 @@ export function important(
...
             break
           }
-          if (isQuote(char)) {
-            isQuoteAfterCandidate = true
-            break
-          }
+          currentLineAfterCandidate += char
         }
 
+        // Heuristics 1: Require the candidate to be inside quotes
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/template/codemods/important.ts:66]:
```suggestion
        // Heuristic 1: Require the candidate to be inside quotes
```

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts [hunk 4]
```diff
@@ -34,32 +46,54 @@ export function important(
...
+        )
+        let isQuoteAfterCandidate = QUOTES.some((quote) =>
+          currentLineAfterCandidate.includes(quote),
+        )
         if (!isQuoteBeforeCandidate || !isQuoteAfterCandidate) {
-          continue
+          continue nextCandidate
+        }
+
+        // Heuristics 2: Disallow object access immediately following the candidate
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/template/codemods/important.ts:77]:
```suggestion
        // Heuristic 2: Disallow object access immediately following the candidate
```

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts [hunk 5]
```diff
@@ -34,32 +46,54 @@ export function important(
...
+        if (currentLineAfterCandidate[0] === '.') {
+          continue nextCandidate
+        }
+
+        // Heuristics 3: Disallow logical operators proceeding or following the candidate
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/template/codemods/important.ts:82]:
```suggestion
        // Heuristic 3: Disallow logical operators preceding or following the candidate
```

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts [hunk 6]
```diff
@@ -34,32 +46,54 @@ export function important(
...
+        for (let operator of LOGICAL_OPERATORS) {
+          if (
+            currentLineAfterCandidate.trim().startsWith(operator) ||
+            currentLineBeforeCandidate.trim().endsWith(operator)
+          ) {
+            continue nextCandidate
+          }
+        }
+
+        // Heuristics 4: Disallow conditional template syntax
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/template/codemods/important.ts:92]:
```suggestion
        // Heuristic 4: Disallow conditional template syntax
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -7,7 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-- Nothing yet!
+### Fixed
+
+- Don't migrate important modifiers when these are used inside conditionals of certain template languages (e.g. `<div v-if="!border"/>`) ([#14774](https://github.com/tailwindlabs/tailwindcss/pull/14774))
```
Comment (adamwathan) [inline CHANGELOG.md:12]:
```suggestion
- Don't migrate important modifiers inside conditional statements in Vue and Alpine (e.g. `<div v-if="!border" />`) ([#14774](https://github.com/tailwindlabs/tailwindcss/pull/14774))
```

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts (hunk 1 above)
Comment (philipp-spiess) [reply packages/@tailwindcss-upgrade/src/template/codemods/important.ts:11]: Ah, I only looked at `v-else` which takes no arguments but `v-else-if` is legit.

File: packages/@tailwindcss-upgrade/src/template/codemods/important.ts (hunk 2 above)
Comment (philipp-spiess) [reply packages/@tailwindcss-upgrade/src/template/codemods/important.ts]: Sorted alphabetically now

**Review by adamwathan — APPROVED** [review]

**Pull request merged**